### PR TITLE
Expire user password

### PR DESCRIPTION
**Pull request by HariBhandari07**
### Description
This PR adds acceptance tests for user password expiration

### Related Issue
https://github.com/owncloud/password_policy/issues/126

### CODE REVIEW

File: tests/acceptance/features/cliPasswordExpire/occExpireUserPassword.feature [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    Then as "Alice" file "/textfile0.txt" should not exist
+#    Then as "Alice" file "/textfile0.txt" should exist
+
+  Scenario: admin expires password of a user and then the user tries to share a file
+    Given using new DAV path
+    Given user "Brian" has been created with default attributes and skeleton files
+    And the administrator has enabled the days until user password expires user password policy
+    And the administrator has expired the password of user "Alice"
+#    Then user "Alice" should not be able to share file "welcome.txt" with user "Brian" using the sharing API
+    When user "Alice" shares file "/welcome.txt" with user "Brian" using the sharing API
```
Comment (HariBhandari07) [inline tests/acceptance/features/cliPasswordExpire/occExpireUserPassword.feature:43]: @phil-davis  in this case,  response xml is empty so both "should" and "should not" step fails with ` Received empty response where XML was expected (Exception)`. Should I create a new sharing step? OR should I create an issue about empty response?

File: tests/acceptance/features/cliPasswordExpire/occExpireUserPassword.feature [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    When user "Alice" deletes file "/textfile0.txt" using the WebDAV API
+    Then as "Alice" file "/textfile0.txt" should not exist
+#    Then as "Alice" file "/textfile0.txt" should exist
+
+  Scenario: admin expires password of a user and then the user tries to share a file
+    Given user "Brian" has been created with default attributes and skeleton files
+    And the administrator has enabled the days until user password expires user password policy
+    And the administrator has expired the password of user "Alice"
+#    Then user "Alice" should not be able to share file "welcome.txt" with user "Brian" using the sharing API
+    When user "Alice" shares file "/welcome.txt" with user "Brian" using the sharing API
```
Comment (HariBhandari07) [inline tests/acceptance/features/cliPasswordExpire/occExpireUserPassword.feature:42]: @phil-davis  what should I do about this fail, as the response is empty?

File: tests/acceptance/features/lib/PasswordPolicySettingsPage.php [hunk 3]
```diff
@@ -67,6 +67,11 @@ class PasswordPolicySettingsPage extends OwncloudPage {
 	];
 	private $passwordPolicyFormId = 'password_policy';
 	private $saveButtonValue = 'Save';
+	protected $currentPasswordInputId = "current_password";
+	protected $newPasswordInputId = "new_password";
+	protected $confirmNewPasswordInputId = "confirm_password";
+	protected $chooseNewPasswordFormXpath = "//form[@id='password_policy']";
+	protected $submitLoginId = "submit";
```
Comment (Talank) [inline tests/acceptance/features/lib/PasswordPolicySettingsPage.php:74]: Why protected? :thinking:

**Review by swoichha — COMMENTED** [review]
LGTM

File: tests/acceptance/features/bootstrap/WebUIPasswordPolicyContext.php [hunk 4]
```diff
@@ -170,6 +176,53 @@ public function theAdminSetsLowercaseLettersUsingTheWebui(
...
+	 *
+	 * @return OwncloudPage
+	 * @throws \Exception
+	 */
+	public function userLogsInWithExpiredPasswordUsingTheWebui($username) {
+		$usernameActual = $this->featureContext->getActualUsername($username);
+		$this->webUILoginContext->theUserBrowsesToTheLoginPage();
+		$this->loginWithExpiredPassword(
+			$usernameActual,
+			$this->featureContext->getPasswordForUser($usernameActual)
```
Comment (kiranparajuli589) [inline tests/acceptance/features/bootstrap/WebUIPasswordPolicyContext.php:222]:
```suggestion
			$this->featureContext->getPasswordForUser($username)
```
The actual username is fetched inside this function.

File: tests/acceptance/features/bootstrap/WebUIPasswordPolicyContext.php [hunk 5]
```diff
@@ -170,6 +176,53 @@ public function theAdminSetsLowercaseLettersUsingTheWebui(
...
+	 */
+	public function loginWithExpiredPassword($username, $password) {
+		$session = $this->getSession();
+		$this->loginPage->waitTillPageIsLoaded($session);
+		$this->loginPage->loginAs(
+			$username,
+			$password
+		);
+		$this->featureContext->asUser($username);
+	}
```
[the remark this answers is not in the document]
Comment (HariBhandari07) [reply tests/acceptance/features/bootstrap/WebUIPasswordPolicyContext.php:207]: Yes, it's a `loginPage` PO function. Login with expired password doesn't occur in other normal cases. So, @phil-davis  and I had a discussion and we decided to put it in `passwordPolicyContext`.

File: tests/acceptance/features/bootstrap/PasswordPolicyContext.php [hunk 6]
```diff
@@ -360,6 +360,19 @@ public function theAdminTogglesForcePasswordChangeOnFirstLoginPasswordPolicyUsin
...
+	 * @Given the administrator has expired the password of user :username
+	 * @When the administrator expires the password of user :username using the occ command
+	 *
+	 * @param string $username
+	 *
+	 * @return void
+	 * @throws Exception
+	 */
+	public function theAdministratorHasExpiredThePasswordOfUser($username) {
+		$this->expireUserPassword($username);
```
[the remark this answers is not in the document]
Comment (HariBhandari07) [reply tests/acceptance/features/bootstrap/PasswordPolicyContext.php:373]: Yeah, I forgot you did a huge refactoring splitting Given-When :laughing: :rofl:

File: tests/acceptance/features/bootstrap/WebUIPasswordPolicyContext.php (hunk 4 above)
Comment (HariBhandari07) [reply tests/acceptance/features/bootstrap/WebUIPasswordPolicyContext.php:222]: @kiranparajuli589 can you elaborate more? Here, I am getting password for a user using the `actual name`.

File: tests/acceptance/features/bootstrap/PasswordPolicyContext.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (HariBhandari07) [reply tests/acceptance/features/bootstrap/PasswordPolicyContext.php:391]: It's done

**Comment by kiranparajuli589** [comment]
LGTM :+1:

**Review by kiranparajuli589 — COMMENTED** [review]
LGTM :+1:

**Review by phil-davis — APPROVED** [review]

**Pull request merged**